src/main/scala/uoption/package.scala

```diff
@@ -0,0 +1,73 @@
+import scala.language.higherKinds
+
+package object uoption {
+  private[uoption] final class WrappedNone(
+      depth: Int,
+      val unwrap: Any // UNone or WrappedNone
+  ) {
+    lazy val wrap: WrappedNone =
+      new WrappedNone(depth + 1, this)
+
+    private val stringRepr: String =
+      ("USome(" * depth) + "UNone" + (")" * depth)
+
+    override def toString(): String = stringRepr
+  }
+
+  object UNone {
+    private[uoption] val wrap: WrappedNone = new WrappedNone(1, this)
+
+    override def toString(): String = "UNone"
+  }
+
+  type UOption[+A] >: UNone.type <: AnyRef
+
+  object UOption {
+    @inline // only for Scala.js?
+    def apply[A](x: A): UOption[A] =
+      if (x == null) UNone
+      else USome(x)
+  }
+
+  type USome[+A] <: UOption[A]
+
+  object USome {
+    @inline // only for Scala.js?
+    def apply[A](value: A): USome[A] = value match {
+      case value @ UNone      => value.wrap.asInstanceOf[USome[A]]
+      case value: WrappedNone => value.wrap.asInstanceOf[USome[A]]
+      case _                  => value.asInstanceOf[USome[A]]
+    }
+
+    // UOptionOps fits the contract of name-based pattern matching
+    def unapply[A](option: UOption[A]): UOptionOps[A] =
+      new UOptionOps(option)
+  }
+
+  implicit class UOptionOps[A](private val self: UOption[A]) extends AnyVal {
+    @inline def isEmpty: Boolean = self eq UNone
+    @inline def isDefined: Boolean = !isEmpty
+
+    /** Must not be called when `isEmpty` is `true`! */
+    @inline // only for Scala.js?
+    private def forceGet: A = (self: Any) match {
+      case none: WrappedNone =>
+        none.unwrap.asInstanceOf[A]
+      case _ =>
+        self.asInstanceOf[A]
```

Comment:
So, IIUC, I can't distinguish `USome(USome(x)): Any` and `USome(USome(USome(x))): Any` (unlike with `Option`), because the number of wrappings is only counted by the type. A `HashMap[Key, Any]` is an uncommon but semi-natural use-case for upcasting values to `Any` and trying to match them back. If I wanted to use such a map, I'd be fine with having to convert into `Option` or something.

That's a good tradeoff and I'd use it nonetheless, I just guess a production library might want to document this to not confuse users.

Or maybe I missed something?

Reply:
That's right. In fact you can't even distinguish `x: Any` and `USome(x): Any`. They are `eq`.